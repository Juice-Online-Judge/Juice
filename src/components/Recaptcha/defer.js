import pTap from 'p-tap'

const defer = () => {
  let state = false
  const deferred = {
    resolved() {
      return state
    }
  }

  deferred.promise = new Promise(resolve => {
    deferred.resolve = resolve
  }).then(
    pTap(() => {
      state = true
    })
  )

  return deferred
}

export default defer
