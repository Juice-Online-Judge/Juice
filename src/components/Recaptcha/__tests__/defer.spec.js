import defer from '../defer'

describe('defer', () => {
  it('Return defer object', () => {
    const deferred = defer()
    expect(deferred).toMatchSnapshot()
  })

  describe('#resolve', () => {
    it('Resolve with given value', async () => {
      const deferred = defer()
      const value = 42

      deferred.resolve(value)
      expect(await deferred.promise).toBe(value)
    })
  })

  describe('#resolved', () => {
    it('Return state', done => {
      const deferred = defer()

      expect(deferred.resolved()).toBe(false)
      deferred.resolve()
      setImmediate(() => {
        expect(deferred.resolved()).toBe(true)
        done()
      })
    })
  })
})
