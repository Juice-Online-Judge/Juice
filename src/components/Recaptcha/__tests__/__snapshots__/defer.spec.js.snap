// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defer Return defer object 1`] = `
Object {
  "promise": Promise {
    "_then": Array [],
  },
  "resolve": [Function],
  "resolved": [Function],
}
`;
